lib/gas_adjuster/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+/// It is used to adjust the base_fee of transactions sent to L1.
+#[derive(Debug)]
+pub struct GasAdjuster {
+    base_fee_statistics: GasStatistics<u128>,
+    blob_base_fee_statistics: GasStatistics<u128>,
+
+    config: GasAdjusterConfig,
+    provider: DynProvider,
+    pubdata_price_sender: watch::Sender<Option<u128>>,
+}
+
+#[derive(Debug, Clone)]
+pub enum PubdataMode {
+    Blobs,
+    Calldata,
+    Validium,
+}
+
+#[derive(Debug)]
+pub struct GasAdjusterConfig {
+    pub pubdata_mode: PubdataMode,
+    pub max_base_fee_samples: usize,
+    pub num_samples_for_blob_base_fee_estimate: usize,
+    pub max_priority_fee_per_gas: u128,
+    pub poll_period: Duration,
+    pub l1_gas_pricing_multiplier: f64,
+    pub pubdata_pricing_multiplier: f64,
+}
+
+impl GasAdjuster {
+    pub async fn new(
+        provider: DynProvider,
+        config: GasAdjusterConfig,
+        pubdata_price_sender: watch::Sender<Option<u128>>,
+    ) -> anyhow::Result<Self> {
+        // Subtracting 1 from the "latest" block number to prevent errors in case
+        // the info about the latest block is not yet present on the node.
+        // This sometimes happens on Infura.
+        let current_block = provider.get_block_number().await?.saturating_sub(1);
+        let fee_history =
+            Self::base_fee_history(&provider, current_block, config.max_base_fee_samples as u64)
+                .await?;
+
+        let base_fee_statistics = GasStatistics::new(
+            config.max_base_fee_samples,
+            current_block,
+            fee_history.iter().map(|fee| fee.base_fee_per_gas),
+        );
+
+        let blob_base_fee_statistics = GasStatistics::new(
+            config.num_samples_for_blob_base_fee_estimate,
+            current_block,
+            fee_history.iter().map(|fee| fee.base_fee_per_blob_gas),
+        );
+
+        let this = Self {
+            base_fee_statistics,
+            blob_base_fee_statistics,
+            config,
+            provider,
+            pubdata_price_sender,
+        };
+        this.pubdata_price_sender
+            .send_replace(Some(this.pubdata_price_inner()));
+
+        Ok(this)
+    }
+
+    /// Performs an actualization routine for `GasAdjuster`.
+    /// This method is intended to be invoked periodically.
+    pub async fn keep_updated(&self) -> anyhow::Result<()> {
+        // Subtracting 1 from the "latest" block number to prevent errors in case
+        // the info about the latest block is not yet present on the node.
+        // This sometimes happens on Infura.
+        let current_block = self.provider.get_block_number().await?.saturating_sub(1);
+
+        let last_processed_block = self.base_fee_statistics.last_processed_block();
+
+        if current_block > last_processed_block {
+            let n_blocks = current_block - last_processed_block;
+            let fee_data = Self::base_fee_history(&self.provider, current_block, n_blocks).await?;
```

Comment:
Given that you already have to provide a field from `&self`, maybe it makes sense to add the receiver type to the method? Keeping it static doesn't seem to bring value tbh.